### DISCUSSION TITLE
Poor performance when many atoms derive value from a single atom

### DISCUSSION THREAD

**Discussion by jdefting** [pmndrs/jotai#2368] [General]
In my use-case, I have ~1000-5000 elements which all watch a global state value. Only a small subset (1-10) of these elements need to re-render anytime the global state changes. 

I would like to implement this with atoms which are derived from the global state atom. My implementation is something like the following (simplified for example).

```ts
const globalAtom = atom(0);
// watch global atom
const derivedAtoms = elements.map((_, i) => atom((get) => get(globalAtom) === i)));
```

However, the performance I get from this is not great, especially compared to using `React.memo()`. 

```tsx
const MemoElement = memo(({ isSelected }) => ...);

const Container = () => {
  const [value, setValue] = useState(0);
  return <div>{elements.map((_, i) => <MemoElement isSelected={value === i} key={i} />)}</div>  
}
```

Example: https://codesandbox.io/p/sandbox/polished-https-45x35t?file=%2Fsrc%2FApp.js%3A10%2C19

Is this performing as expected? This performance delta is surprising to me because Jotai is doing all its calculation outside of React. I assume Jotai is basically looping through the list of subscribers, executing the derivation function, then doing a diff. Conceptually, this doesn't seem like it should perform worse than when React does diff checks with `memo()`.

**Comment by dai-shi** [comment]
Not sure if it's related, but does #2363 help?
https://ci.codesandbox.io/status/pmndrs/jotai/pr/2363
☝️ Find "Local Install Instructions"

**Reply by jdefting** [reply]
Yes, it helps significantly. However, performance does still lag behind `React.memo()` at a higher number of elements and after prod build (not running in dev mode). 

<details><summary>n=10000, Jotai</summary>

![image](https://github.com/pmndrs/jotai/assets/87446002/dafbc975-2140-43bb-9373-6c91dc6355ac)

</details> 

<details><summary>n=10000, React.memo()</summary>

![image](https://github.com/pmndrs/jotai/assets/87446002/6b334f3e-afd0-4fa0-b84a-f6c4673599f3)

</details>

**Reply by dai-shi** [reply]
That's great to hear. I'm not sure what else we can do without investigating further where's the bottleneck. Would you be able to dig into it?

cc who might be interested in it: @edkimmel @samkline @cltnschlosser

**Reply by cltnschlosser** [reply]
Without #2363, it's almost all in recomputeDependents, which is the optimized area
<img width="1675" alt="image" src="https://github.com/pmndrs/jotai/assets/1498529/88dbce1f-83e9-4c3c-8790-d8578697b07f">

With #2363: 
<img width="1593" alt="image" src="https://github.com/pmndrs/jotai/assets/1498529/0bf1d1a1-2f46-42dc-9184-8f047eb5bbc2">

EDIT: Sorry bad screenshots, first one was 93ms and the second was 3.6ms for the onClick function. And these tests were done with your shared codesandbox (2000 atoms)

(note, I'm using the commit prior to this one, as it seems to have broken the functionality, I'll see if I can create a unit test that catches the issue: https://github.com/pmndrs/jotai/pull/2363/commits/5693166880e7feb47dcd2685993b95f0881367b9)
NOTE-EDIT: Issue can't be reproduced with a unit test because it's only an issue in the browser with the generated JS code.

**Reply by samkline** [reply]
Nice find @cltnschlosser ! I just updated the PR #2363 with a fix. For those out of the loop, the issue applies when the code is transpiled to support older JS versions. The transpiled code has an issue with using `Array.prototype.push` with the Set iterator, such that the `push` does nothing. It worked fine in modern JS environments. This was the fix: https://github.com/pmndrs/jotai/pull/2363/commits/4cb7b8361e137c1ea58f0f76794641bc5d48f889

**Reply by samkline** [reply]
Anyway, 3ms of overhead for 2,000 atoms or 10ms for 10,000 atoms does seem pretty reasonable to me, and that's scaling (only) linearly, which is good. I doubt there's much improvement left to be had, it's always going to be a _bit_ slower than pure React.

**Reply by cltnschlosser** [reply]
The render phase following the on click should be faster in the jotai case because only 2 cells need to evaluated, where as pure react will evaluate all of them.

That would be more important in a real use case too, where it’s not just single div squares.

**Reply by jdefting** [reply]
> I doubt there's much improvement left to be had, it's always going to be a bit slower than pure React.

For what it's worth, Zustand actually outperforms pure React in this case. https://codesandbox.io/p/sandbox/festive-cloud-xpd255

<details>
<summary>React.memo() (4x CPU slowdown)</summary>

![image](https://github.com/pmndrs/jotai/assets/87446002/24116976-e491-4e01-8040-34db5d391851)

</details>

<details>
<summary>Zustand (4x CPU slowdown)</summary>

![image](https://github.com/pmndrs/jotai/assets/87446002/be420533-f024-4284-a19a-b0fd00e2651f)

</details>

Maybe there's a fundamental difference that allows Zustand to be more performant in this case. Anyway, the improvements that have been made so far are good enough for my use-case.

**Comment by AjaxSolutions** [comment]
Can you use [useGlobalMemo](https://gist.github.com/tannerlinsley/a1c7e118164b0a1e2c14d2952b9248fb) in this case?
